runtime/compiler/env/J9KnownObjectTable.cpp

```diff
@@ -56,28 +65,176 @@ J9::KnownObjectTable::isNull(Index index)
 TR::KnownObjectTable::Index
 J9::KnownObjectTable::getIndex(uintptrj_t objectPointer)
    {
-   TR_J9VMBase *fej9 = (TR_J9VMBase *)(self()->fe());
-   TR_ASSERT(fej9->haveAccess(), "Must haveAccess in J9::KnownObjectTable::getIndex");
    if (objectPointer == 0)
       return 0; // Special Index value for NULL
 
-   // Search for existing matching entry
-   //
    uint32_t nextIndex = self()->getEndIndex();
-   for (uint32_t i = 1; i < nextIndex; i++)
-      if (*_references.element(i) == objectPointer)
-         return i;
-
-   // No luck -- allocate a new one
-   //
-   J9VMThread *thread = getJ9VMThreadFromTR_VM(self()->fe());
-   TR_ASSERT(thread, "assertion failure");
-   _references.setSize(nextIndex+1);
-   _references[nextIndex] = (uintptrj_t*)thread->javaVM->internalVMFunctions->j9jni_createLocalRef((JNIEnv*)thread, (j9object_t)objectPointer);
+#if defined(J9VM_OPT_JITSERVER)
+   if (self()->comp()->isOutOfProcessCompilation())
```

Comment:
I don't think it's safe for the server to ever execute this code because it deals with objects. Any object pointer may have received from the client could become stale because the client cannot hold VMaccess while sending a message to the server.
It may be the case that after all the transformations we actually don't call `getIndex()` from the server, but it's better to put a ASSERT_FATAL here.

Reply:
> On the client side, client->write() is inside the scope of the critical section.

I missed that. If possible we have to take it out, though I am afraid we may have many transgressions like that. Fortunately, the writes are more forgiving because they block only until the kernel receives the message. In some corner cases the kernel may have a big backlog and therefore the wrote could block.

Reply:
`getIndexAt()` and `getExistingIndexAt()` are safe because they don't manipulate objects.
`getIndex()` uses an object, which may have been moved.
`getPointer()` returns an object and it's safe to use, but the usage of the returned object is not safe.

Reply:
I don't see `getIndex` called by the server in openj9. I do see some calls in OMR in VPHandlers.cpp which are bypassed. You also don't see any messages for `getIndex` so we may be good, but we need that ASSERT in `getIndex` to make sure.

Reply:
> though I am afraid we may have many transgressions like that

Yeah, a quick search in `handleServerMessage()`, I noticed many message handlings that require access to VM have `client->write()` inside the scope of `TR::VMAccessCriticalSection()`. 

For this PR, I will take ` client->write()` out of the scope of VM access critical section for the KOT related messages first.